src/Featurizers/Components/WindowFeaturizerBase.h

```diff
@@ -0,0 +1,244 @@
+
+            ++_cur_index;
+            _cur_index %= _size;
+            return *this;
+        }
+
+        // Post-increment operator.
+        CircularIterator operator++ (int) {
+            CircularIterator tmp(*this);
+
+            assert(++_cur_increment <= _max_increments);
+
+            ++_cur_index;
+            _cur_index %= _size;
+            return tmp;
+        }
+
+        // Two-way comparison: v.begin() == v.cbegin() and vice versa.
+        template<class OtherType>
+        bool operator == (const CircularIterator<OtherType>& rhs) const {
+            return _itr == rhs._itr && _cur_index == rhs._cur_index &&
+                !((_cur_increment == _max_increments) ^ (rhs._cur_increment == rhs._max_increments));
+        }
+
+        template<class OtherType>
+        bool operator != (const CircularIterator<OtherType>& rhs) const {
+            return !operator==(rhs);
+        }
+
+        T& operator* () const {
+            assert(_cur_increment != _max_increments);
+            return *(_itr+_cur_index);
+        }
+
+        T* operator-> () const {
+            assert(_cur_increment != _max_increments);
+            return (_itr+_cur_index);
+        }
+
+        // One way conversion: iterator -> const_iterator
+        operator CircularIterator<const T>() const {
+            return CircularIterator<const T>(_itr);
+        }
+
+    };
+
+    /////////////////////////////////////////////////////////////////////////
+    ///  \class         CircularBuffer
+    ///  \brief         A custom container class created for operations with shifted windows
+    ///                 The goal is to minimize the memory allocation so when the item limit
+    ///                 has been reached it overwrites the oldest item it is storing without
+    ///                 allocating new memory.
+    ///
+    template <class T>
+    class CircularBuffer {
+
+        // Maximum number of elements this container can hold. When this number is reached
+        // the oldest value is overwritten.
+        size_t                                  _max_size;
+
+        // The start position of this circular buffer. This class is implemented by wrapping around
+        // a vector so the "start" will change based on how much data has been overwritten.
+        size_t                                  _start_offset;
+
+        // Current number of element held by this container. It start from 0 and grows to _max_size
+        size_t                                  _cur_size;
+
+        // The vector that holds the actual data. This is set to hold _max_size elements and then
+        // is never resized.
+        std::vector<T>                          _data;
+
+    public:
+        // ----------------------------------------------------------------------
+        // |  Public Data
+        typedef CircularIterator<T>             iterator;
+        typedef CircularIterator<const T>       const_iterator;
```

Comment:
Prefer using statements over typedefs.

Reply:
We did the typedef cause in your email to us you said we needed to provide typedefs. Would `using` be better? Or did we misunderstand and put the typedefs in the wrong spot?

Reply:
My bad. I am using old terminology with new concepts. I should have said "we should have alias types for iterator and const_iterator". There are 2 ways of creating alias types in C++ - 1) via typedef (old with corner cases) 2) via using (new without corner cases).

---
In reply to: [385338421](https://github.com/microsoft/FeaturizersLibrary/pull/154#discussion_r385338421) [](ancestors = 385338421)

Reply:
Replaced the `typedefs` with `using`